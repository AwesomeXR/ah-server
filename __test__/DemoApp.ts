import {
  createApp,
  BaseController,
  IContext,
  BaseService,
  IMiddleware,
  router,
  lifeCycle,
} from '../src';

class InspectService extends BaseService {
  @lifeCycle('setup')
  protected setup() {
    this.logger.info('InspectService init');
  }

  @lifeCycle('listen')
  protected listen() {
    this.logger.info('InspectService listen');
  }

  @lifeCycle('run')
  protected run() {
    this.logger.info('InspectService run');
  }

  @lifeCycle('close')
  protected close() {
    this.logger.info('InspectService close');
  }

  echo(text: string) {
    return text;
  }
}

class EchoController extends BaseController {
  @router({
    path: '/echo',
    method: ['GET', 'POST'],
    query: {
      schema: {
        type: 'object',
        properties: { text: { type: 'string' } },
      },
    },
  })
  async echo(_ctx: IContext, q: { text?: string }) {
    const output = (this.service as any).inspect.echo(q.text);
    await new Promise(resolve => setTimeout(resolve, 500));
    return { output };
  }
}

class HomeController extends BaseController {
  @router({ path: '/', method: ['GET'] })
  async index(ctx: IContext) {
    ctx.body = '<h1>Hi</h1>';
  }

  @router({
    path: '/file',
    method: ['POST'],
    query: {
      schema: {
        type: 'object',
        properties: {
          a: { type: 'string' },
          file: {
            type: 'object',
            properties: { path: { type: 'string' } },
            required: ['path'],
          },
        },
        required: ['a'],
      },
    },
  })
  async uploadFile(_ctx: IContext, q: { a: string; file: { path: string } }) {
    return q;
  }

  @router({
    path: '/queryTap',
    method: ['GET'],
    query: {
      tap: 'tryParseIntProperty',
      schema: {
        type: 'object',
        properties: {
          title: { type: 'string' },
          pageNum: { type: 'integer' },
        },
        required: ['title', 'pageNum'],
      },
    },
  })
  async queryTap(_ctx: IContext, q: { pageNum: number; title: string }) {
    return q;
  }
}

const testMiddlewareA: IMiddleware = async (ctx, next) => {
  ctx.response.set({ 'x-a': 'a' });
  return next();
};

const testMiddlewareB: IMiddleware = async (ctx, next) => {
  ctx.response.set({ 'x-b': 'b' });
  return next();
};

export const createTestApp = (config: any) => {
  return createApp({
    config,
    service: { inspect: InspectService },
    controllers: [HomeController, EchoController],
    middlewares: [testMiddlewareA, testMiddlewareB],
  });
};
