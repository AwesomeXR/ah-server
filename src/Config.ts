declare module '.' {
  interface IConfig extends BaseConfig {}
}

export class BaseConfig {
  readonly LOCAL_PORT: number = +(process.env.AH_LOCAL_PORT || process.env.LOCAL_PORT || '10001');
  readonly HOSTNAME = process.env.AH_HOSTNAME || process.env.HOSTNAME || 'localhost';

  readonly HTTPS_KEY = process.env.AH_HTTPS_KEY || process.env.HTTPS_KEY;
  readonly HTTPS_CERT = process.env.AH_HTTPS_CERT || process.env.HTTPS_CERT;
  readonly AUTH_SALT = process.env.AH_AUTH_SALT || process.env.AUTH_SALT || 'x';

  sequelize() {
    return Object.entries(this)
      .filter(([, v]) => typeof v !== 'undefined')
      .map(([n, v]) => `${n}=${v}`)
      .join(' ');
  }
}
