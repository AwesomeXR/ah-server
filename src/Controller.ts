import { IApplication } from '.';
import { BaseService } from './Service';

export class BaseController extends BaseService {}

export interface IBaseController extends BaseController {}
export interface IBaseControllerCls {
  new (app: IApplication): IBaseController;
}
