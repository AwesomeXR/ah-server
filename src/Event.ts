import * as http from 'http';

export const ReadyEvt = Symbol('ReadyEvt');
export interface ReadyEvt {
  server: http.Server;
}

export const CloseEvt = Symbol('CloseEvt');

export const ErrorEvt = Symbol('ErrorEvt');
export interface ErrorEvt {
  error: Error;
}
