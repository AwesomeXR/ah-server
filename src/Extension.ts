import { ILifeCycleName, ILifeCycleHook, IBaseServiceCls, IMiddleware } from '.';

export type IBaseExtension = {
  app?: Record<string, any>;
  service?: Record<string, IBaseServiceCls>;
  middleware?: IMiddleware[];
  lifeCycle?: Partial<Record<ILifeCycleName, ILifeCycleHook>>;
};
