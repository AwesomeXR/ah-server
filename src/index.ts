import 'reflect-metadata';

// for type
import * as Koa from 'koa';
import 'koa-body';

import { Schema } from 'ah-api-type';

export interface IApplication {}
export type ILifeCycleName = 'setup' | 'listen' | 'run' | 'close';
export type ILifeCycleHook = (app: IApplication) => Promise<any>;

export interface IService {}
export interface IContext extends Koa.Context {}

export type IRouterMethod = 'GET' | 'POST' | 'DELETE' | 'PUT';
export type IMiddleware = (ctx: IContext, next: () => Promise<any>) => Promise<any>;
export interface IRouterMeta {
  path: string;
  method: IRouterMethod | IRouterMethod[];
  middlewares?: IMiddleware[];
  query?: {
    tap?: ((q: any) => any) | 'tryParseIntProperty';
    schema: Schema;
  };
}

export interface IConfig {
  LOCAL_PORT: number;
}

export * from './App';
export * from './Config';
export * from './Controller';
export * from './Service';
export * from './util';
export * from './Scheduler';
export * from './error';
export * from './Event';
export * from './Extension';
export * from './refactor';
