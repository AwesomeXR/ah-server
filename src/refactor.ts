import { ILifeCycleName, IRouterMeta } from '.';

export const MetaKey = {
  route: 'ah:design:route',
  lifeCycle: 'ah:design:lifeCycle',
};

/** 路由注解 */
export const router = (meta: IRouterMeta) => Reflect.metadata(MetaKey.route, meta);
export const getRouterMeta = (target: any, property: string): IRouterMeta | undefined =>
  Reflect.getMetadata(MetaKey.route, target, property);

/** 生命周期注解 */
export const lifeCycle = (name: ILifeCycleName) => Reflect.metadata(MetaKey.lifeCycle, name);
export const getLifeCycleMeta = (target: any, property: string): ILifeCycleName | undefined =>
  Reflect.getMetadata(MetaKey.lifeCycle, target, property);
