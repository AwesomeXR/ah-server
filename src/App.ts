import { CronJob } from 'cron';
import * as Koa from 'koa';
import * as Router from 'koa-router';
import * as koaBody from 'koa-body';
import * as urllib from 'urllib';
import { Logger } from 'ah-logger';
import { BaseScheduler } from './Scheduler';
import { getLifeCycleMeta, IApplication, IBaseSchedulerCls, IBaseServiceCls, IConfig, IContext, ILifeCycleName, IMiddleware } from '.';
import { ErrorTypeEnum } from './error';
import { getOwnPropertyEntries, mapValues, pick, tryParseIntProperty, validate } from './util';
import { BaseController, IBaseControllerCls } from './Controller';
import * as http from 'http';
import * as https from 'https';
import { CloseEvt, ErrorEvt, ReadyEvt } from './Event';
import * as fs from 'fs';
import { IBaseExtension } from './Extension';
import { getRouterMeta } from './refactor';
import * as jwt from 'jsonwebtoken';
import { v4 as uuid } from 'uuid';
import * as crypto from 'crypto';

declare module '.' {
  // eslint-disable-next-line
  interface IApplication extends Koa {
    config: IConfig;
    logger: Logger;
    service: IService;
    controllers: BaseController[];
    schedulers: BaseScheduler[];
    middlewares: IMiddleware[];

    lifeCycleHooks: Record<ILifeCycleName, ILifeCycleHook[]>;

    curl: <T>(url: string, opt?: urllib.RequestOptions | undefined) => Promise<urllib.HttpClientResponse<T>>;
    runInBackground: (callback: () => Promise<any>) => void;
    jwt: {
      sign: <T extends Record<string, any>>(payload: T, opt?: jwt.SignOptions) => string;
      verify: <T extends unknown>(token: string, opt?: jwt.VerifyOptions | undefined) => T | null;
    };
    uuid: (prefix?: string) => string;
    encrypt(input: string): string;
    randomString(length: number): string;
    md5(body: string | Buffer): string;
  }

  // eslint-disable-next-line
  interface IContext {
    validate: typeof validate;
    app: IApplication;
  }
}

export type createAppOpt = {
  config: IConfig;
  service: Record<string, IBaseServiceCls>;
  controllers: IBaseControllerCls[];
  schedulers?: IBaseSchedulerCls[];
  middlewares?: IMiddleware[];
  extensions?: IBaseExtension[];
  lifeCycleHooks?: Partial<IApplication['lifeCycleHooks']>;

  logger?: Logger;

  mixin?: Record<string, any>;
};

export const createApp = ({
  config,
  service,
  controllers,
  schedulers = [],
  middlewares = [],
  extensions = [],
  lifeCycleHooks = {},
  logger = new Logger('APP'),
  mixin,
}: createAppOpt) => {
  const app = new Koa() as IApplication;

  // mixin 要首先执行
  Object.assign(app, mixin);

  // merge base props
  // =================================

  app.config = config;
  app.logger = logger;
  app.service = mapValues(service, S => new S(app));

  app.controllers = controllers.map(C => new C(app));
  app.schedulers = schedulers.map(S => new S(app));

  app.middlewares = middlewares;

  app.lifeCycleHooks = {
    setup: lifeCycleHooks.setup?.concat() || [],
    listen: lifeCycleHooks.listen?.concat() || [],
    run: lifeCycleHooks.run?.concat() || [],
    close: lifeCycleHooks.close?.concat() || [],
  };

  // 扩展 ctx
  Object.assign(app.context, { validate, app });

  // 全局错误
  app.on('error', err => {
    let msg = err.message || err;
    if (err.stack) msg += '\n' + err.stack;

    app.logger.error(msg);

    app.emit(ErrorEvt, { error: err } as ErrorEvt);
  });

  // merge extensions
  // ==============================

  extensions.forEach(ext => {
    if (ext.app) {
      Object.entries(ext.app).forEach(([k, v]) => {
        (app as any)[k] = typeof v === 'function' ? v.bind(app) : v;
      });
    }

    if (ext.service) {
      Object.entries(ext.service).forEach(([k, S]) => {
        (app.service as any)[k] = new S(app);
      });
    }

    if (ext.middleware) ext.middleware.forEach(mid => app.middlewares.push(mid));

    if (ext.lifeCycle) {
      Object.entries(ext.lifeCycle).forEach(([_type, hook]) => {
        const type = _type as ILifeCycleName;
        app.lifeCycleHooks[type].push(hook);
      });
    }
  });

  // 要等上面 extensions 的 middlewares 也 push 进来
  app.middlewares.forEach(m => app.use(m));

  // 注解收集 lifeCycle hooks
  {
    Object.entries(app.service).forEach(([sName, sIns]) => {
      getOwnPropertyEntries(sIns).forEach(([sMethodName, sMethod]) => {
        if (typeof sMethod !== 'function') return;

        const lifeCycleName = getLifeCycleMeta(sIns, sMethodName);
        if (!lifeCycleName) return;

        app.lifeCycleHooks[lifeCycleName].push(async _app => (_app.service as any)[sName][sMethodName]());

        app.logger.info(`register lifeCycle ${lifeCycleName}: ${sName}.${sMethodName}`);
      });
    });
  }

  // setup controller
  {
    app.use(koaBody({ multipart: true }));

    // 构造 router
    const router = new Router<any, IContext>();

    app.controllers.forEach(ctrlIns => {
      getOwnPropertyEntries(ctrlIns).forEach(([pName, handler]) => {
        if (typeof handler !== 'function') return;

        const rMeta = getRouterMeta(ctrlIns, pName);
        if (!rMeta) return;

        app.logger.info(`register controller: ${rMeta.method} ${rMeta.path} -> ${ctrlIns.name}.${handler.name}`);

        const name = [ctrlIns.name, handler.name].join('.');
        const methods = Array.isArray(rMeta.method) ? rMeta.method : [rMeta.method];

        const wrapperMid: IMiddleware = async (ctx, next) => {
          try {
            return await next();
          } catch (err: any) {
            // 自定义异常
            if (Object.values(ErrorTypeEnum).includes(err.type)) {
              ctx.status = err.status;
              ctx.body = pick(err, ['message', 'type', 'code', 'status']);
              return;
            }

            // 其他异常外抛
            throw err;
          }
        };

        const ctrlMids = [
          wrapperMid,
          ...(rMeta.middlewares || []),
          async (ctx: IContext) => {
            let q: any;

            if (rMeta.query) {
              q = {
                ...ctx.params,
                ...ctx.request.query,
                ...ctx.request.body,
                ...ctx.request.files,
              };

              if (rMeta.query.tap) {
                if (typeof rMeta.query.tap === 'function') q = rMeta.query.tap(q);
                else if (rMeta.query.tap === 'tryParseIntProperty') q = tryParseIntProperty(q);
              }

              q = ctx.validate<any>(q, rMeta.query.schema);
            }

            const data = await handler.call(ctrlIns, ctx, q);
            if (data) ctx.body = data;
          },
        ];

        if (methods.includes('GET')) router.get(name, rMeta.path, ...ctrlMids);
        if (methods.includes('POST')) router.post(name, rMeta.path, ...ctrlMids);
        if (methods.includes('PUT')) router.put(name, rMeta.path, ...ctrlMids);
        if (methods.includes('DELETE')) router.delete(name, rMeta.path, ...ctrlMids);
      });
    });

    app.use(router.routes());
    app.use(router.allowedMethods());
  }

  // setup scheduler
  const schedulerStartList: (() => void)[] = app.schedulers
    .map(s => {
      const sLogger = app.logger.extend(s.name);

      if (s.timer.type === 'cron') {
        const { cron } = s.timer;
        return () => {
          new CronJob(
            cron,
            () => {
              s.invoke().catch(e => {
                sLogger.error(`${s.name} error: ${e.message || e}`);
              });
            },
            undefined,
            false, // start
            undefined, // timeZone
            undefined, // context
            !!s.immediately // runOnInit
          ).start();
        };
      }

      if (s.timer.type === 'interval') {
        const { interval } = s.timer;
        return () => {
          const invoke = () => {
            s.invoke()
              .then(() => setTimeout(invoke, interval))
              .catch(e => {
                sLogger.error(`${s.name} error: ${e.message || e}`);
                setTimeout(invoke, interval);
              });
          };

          if (s.immediately) invoke();
          else setTimeout(invoke, interval);
        };
      }
    })
    .filter(v => !!v) as any;

  // builtin utils
  // =============================
  app.curl = async <T>(url: string, opt?: urllib.RequestOptions) => {
    return urllib.request<T>(url, opt);
  };

  app.runInBackground = (callback: () => Promise<any>) => {
    const _asyncStack = new Error('').stack || '<null>';

    setTimeout(() => {
      callback().catch(err => {
        app.logger.error('[BG] %s\n--- async ---\n%s', err.message, _asyncStack);
      });
    }, 0);
  };

  app.jwt = {
    sign: <T extends Record<string, any>>(payload: T, opt: jwt.SignOptions = {}) => jwt.sign(payload, app.config.AUTH_SALT, opt),
    //
    verify: <T extends any>(token: string, opt?: jwt.VerifyOptions) => {
      try {
        return jwt.verify(token, app.config.AUTH_SALT, opt) as T;
      } catch (err) {
        if (err instanceof jwt.JsonWebTokenError) return null;
        if (err instanceof jwt.NotBeforeError) return null;
        if (err instanceof jwt.TokenExpiredError) return null;
        throw err;
      }
    },
  };

  app.uuid = (prefix = '') => {
    return prefix + uuid();
  };

  app.encrypt = (input: string) => {
    return crypto.createHmac('sha256', app.config.AUTH_SALT).update(input).digest('hex');
  };

  app.randomString = (length: number): string => {
    return crypto
      .randomBytes(Math.ceil(length / 2))
      .toString('hex')
      .slice(0, length);
  };

  app.md5 = (body: string | Buffer) => {
    return crypto.createHash('md5').update(body).digest('hex');
  };

  // export apis
  // ================================

  const run = async () => {
    app.logger.info(app.config.sequelize());

    // invoke hook:setup
    app.logger.info('[LifeCycle] setup %s', app.lifeCycleHooks.setup.length);
    await Promise.all(app.lifeCycleHooks.setup.map(h => h(app)));

    const callback = app.callback();

    const server =
      app.config.HTTPS_KEY && app.config.HTTPS_CERT
        ? https.createServer(
            {
              key: fs.readFileSync(app.config.HTTPS_KEY, 'utf-8'),
              cert: fs.readFileSync(app.config.HTTPS_CERT, 'utf-8'),
            },
            callback
          )
        : http.createServer(callback);

    const port = app.config.LOCAL_PORT;
    const hostname = app.config.HOSTNAME;

    server.listen(port, hostname);

    // invoke hook:listen
    app.logger.info('[LifeCycle] listen %s', app.lifeCycleHooks.listen.length);
    await Promise.all(app.lifeCycleHooks.listen.map(h => h(app)));

    app.logger.info(`app start at ${hostname}:${port}`);

    // scheduler 放到启动后
    schedulerStartList.forEach(s => s());

    // invoke hook:run
    app.logger.info('[LifeCycle] run %s', app.lifeCycleHooks.run.length);
    await Promise.all(app.lifeCycleHooks.run.map(h => h(app)));

    app.emit(ReadyEvt, { server } as ReadyEvt);

    // return stop functions
    return async () => {
      // invoke hook:close
      app.logger.info('[LifeCycle] close', app.lifeCycleHooks.close.length);
      await Promise.all(app.lifeCycleHooks.close.map(h => h(app)));

      return new Promise<void>((resolve, reject) => {
        // 优雅退出
        // @see https://zhuanlan.zhihu.com/p/275312155?utm_source=wechat_session&utm_medium=social&utm_oi=39191756931072
        server.close(err => {
          if (err) return reject(err);
          return resolve();
        });
      }).finally(() => app.emit(CloseEvt));
    };
  };

  return { app, run };
};
