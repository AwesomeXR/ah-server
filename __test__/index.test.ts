import { HttpClientResponse } from 'urllib';
import { BaseConfig, ErrorEvt, pick } from '../src';
import { createTestApp } from './DemoApp';

const port = 20000;

function inspectResp(resp: HttpClientResponse<any>) {
  return pick(resp, ['data', 'status']);
}

describe('App', () => {
  const config = new BaseConfig();
  const { app, run } = createTestApp(config);
  let cancel: () => Promise<any>;

  app.on(ErrorEvt, err => {
    throw new Error(err.message);
  });

  beforeAll(async () => {
    cancel = await run();
  });
  afterAll(() => cancel());

  it('curl', async () => {
    const r1 = await app.curl<any>(`http://localhost:${port}/echo?text=hi`, {
      dataType: 'json',
    });
    expect(inspectResp(r1)).toMatchSnapshot();
  });

  it('post body', async () => {
    const r2 = await app.curl<any>(`http://localhost:${port}/echo`, {
      method: 'POST',
      contentType: 'json',
      dataType: 'json',
      data: {
        text: 'hi',
      },
    });
    expect(inspectResp(r2)).toMatchSnapshot();
  });

  it('validate 400', async () => {
    const r3 = await app.curl<any>(`http://localhost:${port}/echo?text=100`, {
      dataType: 'json',
    });
    expect(inspectResp(r3)).toMatchSnapshot();
  });

  it('middleware', async () => {
    const r = await app.curl<any>(`http://localhost:${port}/echo?text=hi`, {
      dataType: 'json',
    });
    expect(r.headers['x-a']).toEqual('a');
    expect(r.headers['x-b']).toEqual('b');
  });

  it('upload file', async () => {
    const r = await app.curl<any>(`http://localhost:${port}/file`, {
      dataType: 'json',
      data: { a: 'a' },
      files: [Buffer.from('x')],
    });
    expect(r.data.a).toBeTruthy();
    expect(r.data.file.path).toBeTruthy();
  });

  it('query tap', async () => {
    const r = await app.curl<any>(`http://localhost:${port}/queryTap`, {
      dataAsQueryString: true,
      data: { pageNum: 1, title: 'aaa' },
      dataType: 'json',
    });
    expect(r.data).toEqual({ pageNum: 1, title: 'aaa' });
  });
});
