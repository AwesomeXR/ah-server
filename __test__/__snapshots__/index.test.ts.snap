// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App curl 1`] = `
{
  "data": {
    "output": "hi",
  },
  "status": 200,
}
`;

exports[`App post body 1`] = `
{
  "data": {
    "output": "hi",
  },
  "status": 200,
}
`;

exports[`App validate 400 1`] = `
{
  "data": {
    "output": "100",
  },
  "status": 200,
}
`;
